[red-knot] document test framework

carljm opened the pull request:
This adds documentation for the new test framework.

I also added documentation for the planned design of features we haven't built yet (clearly marked as such), so that this doc can become the sole source of truth for the test framework design (we don't need to refer back to the original internal design document.)

Also fixes a few issues in the test framework implementation that were discovered in writing up the docs.

dhruvmanila commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,430 @@
+    this line.
+- `# error: [invalid-assignment]` requires that the matched diagnostic have the rule code
+    `invalid-assignment`.
+- `# error: "Some text"` requires that the matched diagnostic's full message contain the text `Some text`.
```
Assuming that the quotes are required, can we clarify that part somewhere here?

dhruvmanila commented on crates/red_knot_test/README.md:
````diff
@@ -0,0 +1,430 @@
+x: str = 1  # error: [invalid-assignment]
+```
+
+Or it may be a comment on its own line, in which case it applies to the next non-assertion-comment
+line:
````
I find "non-assertion-comment" a bit confusing because at first I thought it meant a non-assertion _comment_, so maybe the next statement / code line?

dhruvmanila commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,430 @@
+## Multi-file tests
+
+Some tests require multiple files, with imports from one file to another. Multiple fenced code
+blocks represent multiple embedded files. Since files must have unique names, only zero or one files
```
Should it be "only one file can use..." instead of "only zero or one files can use..."?

dhruvmanila commented on crates/red_knot_test/README.md:
`````diff
@@ -0,0 +1,430 @@
+````markdown
+```py
+from b import C
+reveal_type(C)  # Literal[C]
`````
Did you mean?
```suggestion
reveal_type(C)  # revealed: Literal[C]
```

carljm replied:
I would love to do that! But probably not a priority :)

dhruvmanila commented on crates/red_knot_test/README.md:
`````diff
@@ -0,0 +1,430 @@
+```
+````
+
+We may want to also support testing Jupyter notebooks as embedded files using the `json` language.
`````
Or, maybe some kind of start and end indicator that signals that any code blocks between them should be in a single Jupyter Notebook file.

carljm replied:
I guess it depends what we need to test about notebooks. If all we care about is the Python contents of cells, then something like @dhruvmanila suggests makes sense. If we want to write tests in this format that actually test other characteristics of the JSON structure of a notebook, then I think we should embed the JSON directly. It doesn't help readability to have key aspects of the test hidden in a different file.

dhruvmanila replied:
> If we want to write tests in this format that actually test other characteristics of the JSON structure of a notebook

I'm not sure if Red Knot should be testing the notebook JSON structure, I think it should be done as an integration tests which currently there are :)

But yeah, what you said makes sense and we can think about it when it's required.

carljm replied:
Updated the text here to not mention using `json` language and just say exact syntax is TBD.

dhruvmanila commented on crates/red_knot_test/README.md:
`````diff
@@ -0,0 +1,430 @@
+TOML fenced code block:
+
+````markdown
+```toml
+[tool.knot]
+warn-on-any = true
+
`````
````suggestion
```toml
[tool.knot]
warn-on-any = true
```

````

dhruvmanila commented on crates/red_knot_test/README.md:
````diff
@@ -0,0 +1,430 @@
+This file is written to `/src/test.py`, because the default workspace root is `/src/ and the default
+file path is `test.py`:
+
+```py
+reveal_type(I_AM_THE_ONLY_BUILTIN)  # revealed: Literal[1]
+
````
````suggestion
```py
reveal_type(I_AM_THE_ONLY_BUILTIN)  # revealed: Literal[1]
```

````

carljm commented:
> There is circular imports concept in Python, do you plan support it into mdtests?

I'm not sure what extra support mdtests would need for circular imports? It allows you to create whatever Python files you like with whatever contents you like, so you can create and test any circular import scenario you would want.

carljm commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,430 @@
+databases (so each is a from-scratch run of the type checker, with no data persisting from any
+previous test.)
+
+Due to Rust test runner limitations, an entire test suite (Markdown file) is run as a single Rust
+test, so it's not possible to select individual tests within it to run.
```
It's an inherent limitation of `cargo test` that there is no way to dynamically create tests. I think to directly address this limitation we'd have to either write our own test runner, or use a build script to code-generate tests. Neither of those are planned. 

What I think is more likely to indirectly address this problem is that we could add some kind of "focus" indicator that's easy to add and remove from a test file that will cause it to only run that test from that file. Or we could do something similar that works via env var.

AlexWaygood commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,430 @@
+of these can be replaced by real red-knot config options, maybe some of them will be kept long term
+as test-specific options.
+
+We should be able to configure the default workspace root to something other than `/src/` using a
+`workspace-root` config option.
+
+We should be able to add a third-party root using the `third-party-root` config option.
+
+We may want to add additional config options for setting additional search path kinds.
```
```suggestion
Other future planned changes to configuration include:

- We should be able to configure the default workspace root to something other than `/src/` using a
`workspace-root` config option.

- We should be able to add a third-party root using the `third-party-root` config option.

- We may want to add additional config options for setting additional search path kinds.
```

AlexWaygood commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,430 @@
+Relative embedded-file paths are relative to the workspace root, even if it is explicitly set to a
+non-default value using the `workspace-root` config.
+
+### Specifying custom typeshed
```
```suggestion
### Specifying a custom typeshed
```

AlexWaygood commented on crates/red_knot_test/README.md:
````diff
@@ -0,0 +1,430 @@
+configured above as our custom typeshed root:
+
+```py path=/typeshed/stdlib/builtins.pyi
+I_AM_THE_ONLY_BUILTIN = 1
````
scary

AlexWaygood commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,430 @@
+
+The inline comment diagnostic assertions are useful for making quick, readable assertions about
+diagnostics in a particular location. But sometimes we will want to assert on the full diagnostic
+output of a test. Or sometimes (see “incremental tests” below) we will want to assert on diagnostics
```
```suggestion
output of a tested Python snippet. Or sometimes (see “incremental tests” below) we will want to assert on diagnostics
```

carljm replied:
I'll probably use "embedded Python file" here for consistency

AlexWaygood commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,430 @@
+output of a test. Or sometimes (see “incremental tests” below) we will want to assert on diagnostics
+in a file, without impacting the contents of that file by changing a comment in it. In these cases,
+a Python fenced code block in a test could be followed by a fenced code block with language
+`output`; this contains the full diagnostic output for the preceding test file:
```
```suggestion
`output`; this would contain the full diagnostic output for the preceding test file:
```

AlexWaygood commented on crates/red_knot_test/README.md:
````diff
@@ -0,0 +1,430 @@
+reveal_type(x)
+```
+
+This is just an example, not a proposal that red-knot would every actually output diagnostics in
````
```suggestion
This is just an example, not a proposal that red-knot would ever actually output diagnostics in
```

AlexWaygood commented on crates/red_knot_test/README.md:
`````diff
@@ -0,0 +1,430 @@
+````
+
+We will want to build tooling to automatically capture and update these “full diagnostic output”
+blocks, when tests are run in an update-output mode (probably specified by env var.)
`````
```suggestion
blocks, when tests are run in an update-output mode (probably specified by an environment variable.)
```

AlexWaygood commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,430 @@
+We will want to build tooling to automatically capture and update these “full diagnostic output”
+blocks, when tests are run in an update-output mode (probably specified by env var.)
+
+By default, an `output` block specifies diagnostic output for the file `<workspace-root>/test.py`.
```
```suggestion
By default, an `output` block will specify diagnostic output for the file `<workspace-root>/test.py`.
```

AlexWaygood commented on crates/red_knot_test/README.md:
`````diff
@@ -0,0 +1,430 @@
+contents for `test.py` in stage 1, which we don't want to do in this test.)
+````
+
+Any number of stages can be provided in an incremental test. If a stage re-specifies a filename that
`````
```suggestion
It will be possible to provide any number of stages in an incremental test. If a stage re-specifies a filename that
```

AlexWaygood commented on crates/red_knot_test/README.md:
`````diff
@@ -0,0 +1,430 @@
+````
+
+Any number of stages can be provided in an incremental test. If a stage re-specifies a filename that
+was specified in a previous stage (or the initial state), that file is modified. A new filename
`````
```suggestion
was specified in a previous stage (or the initial stage), that file is modified. A new filename
```

AlexWaygood commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,430 @@
+
+Any number of stages can be provided in an incremental test. If a stage re-specifies a filename that
+was specified in a previous stage (or the initial state), that file is modified. A new filename
+appearing for the first time in a new stage will create a new file. To delete a previously-created
```
```suggestion
appearing for the first time in a new stage will create a new file. To delete a previously created
```

AlexWaygood commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,430 @@
+was specified in a previous stage (or the initial state), that file is modified. A new filename
+appearing for the first time in a new stage will create a new file. To delete a previously-created
+file, specify that file with the tag `delete` in its tag string (in this case, it is an error to
+provide non-empty contents.) Any previously-created files that are not re-specified in a later stage
```
Closing parens should come before the period unless it's a full sentence inside the parens:

```suggestion
provide non-empty contents). Any previously-created files that are not re-specified in a later stage
```

carljm commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,430 @@
+after `# revealed:` must match exactly with the displayed form of the revealed type of that
+expression.
+
+The `reveal_type` function can be imported from the `typing` standard library module (or, for older
+Python versions, from the `typing_extensions` pseudo-standard-library module -- it is a third-party
+module, but typeshed, and thus type checkers also, treat it as part of the standard library):
```
Using a footnote for the digression is a good idea, I'll do that.

> I'd wager that 99% of the time you'll want to import it from `typing_extensions` rather than `typign`

I don't think this is true; I think rather that 99% of tests will not import it at all, and the only ones that do will be a few tests that are explicitly testing the handling of imported vs pseudo-builtin `reveal_type`, and of those few tests probably roughly half (i.e. just one or two tests each) will import it each way.

I think the more important consideration here is that the preferred and modern thing to do, if on a Python version where it's possible, is to import from `typing`, so that's what should be featured, and the legacy approach should be a parenthetical.

carljm commented on crates/red_knot_test/README.md:
````diff
@@ -0,0 +1,430 @@
+reveal_type("foo")  # revealed: Literal["foo"]
+```
+
+For convenience, type checkers also pretend that `reveal_type` is a built-in, so that this import is
+not required. Using `reveal_type` without importing it issues a diagnostic warning that it was used
+without importing it, in addition to the diagnostic revealing the type of the expression.
+
+The `# revealed:` assertion must always match a revealed-type diagnostic, and will also match the
+undefined-reveal diagnostic, if present, so it's safe to use `reveal_type` in tests either with or
+without importing it. (Style preference is to usually not import it in tests, unless specifically
+testing something about the behavior of importing it.)
````
I think the current text accurately and precisely describes how they are handled, and this information is relevant to authors/readers of tests, because the details of how they are handled can be visible to authors of tests (for example, if you do `rt = return_type` and then `rt(foo) # revealed: ...` on another line, you will see unimported-reveal diagnostic on the `rt = return_type` line, and it won't be matched by the `# revealed: ...` assertion. And we will have tests that depend on these distinctions (the tests of the undefined-reveal diagnostic itself.)

The proposed new text isn't noticeably shorter, and it has the disadvantage of being wrong :) It's simply not true that "this diagnostic is never emitted" in tests using mdtest.

carljm replied:
Reading this again, I do think the text could be re-ordered a bit to emphasize the most important thing (`# revealed: ` matches a revealed type from `reveal_type` and then follow up with the stuff about undefined-reveal separately; I'll look at an edit along those lines.

carljm commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,430 @@
+#### error
+
+A comment beginning with `# error:` is an assertion that a type checker diagnostic will
+be emitted, with text span starting on that line. If the comment is simply `# error:`, this will
+match any diagnostic. The matching can be narrowed in three ways:
```
I agree that we should probably never use this, it's just the most natural implementation. Personally I think it's fine to just avoid ever using such assertions, but allow them. If we want to prohibit them, we have to decide what the precise rule should be. Is an assertion that only asserts the column OK? Or should the error code always be required? Or should it be "either error code or message-contains" that is required?

AlexWaygood replied:
> Personally I think it's OK to allow such assertions but just avoid using them.

I think it's inevitable that either one of us or a contributor adds one (either deliberately or accidentally) at some point, if we allow it!

> Or should it be "either error code or message-contains" that is required?

I would say this

carljm replied:
Ok, that seems reasonable, I'll look at making that behavior change in this PR.

carljm commented on crates/red_knot_test/README.md:
`````diff
@@ -0,0 +1,430 @@
+```
+````
+
+Relative file names are always relative to the "workspace root", which is also an import root.
`````
I can try to make it more clear, maybe with reference to `sys.path`.

carljm commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,430 @@
+
+## Multi-test suites
+
+A single test suite (markdown file) can contain multiple tests, by demarcating them using Markdown
```
Agreed, I prefer "Markdown" and will aim to be consistent with that.

carljm commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,430 @@
+databases (so each is a from-scratch run of the type checker, with no data persisting from any
+previous test.)
+
+Due to Rust test runner limitations, an entire test suite (Markdown file) is run as a single Rust
```
I'll probably be more specific here and just say due to `cargo test` limitations

carljm commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,430 @@
+Some tests may need to specify non-Python embedded files: typeshed `stdlib/VERSIONS`, `pth` files,
+`py.typed` files, `pyvenv.cfg` files...
+
+We should allow specifying any of these using the `text` language in the code block tag string:
```
I was trying to steer away from language that suggests an ironclad commitment to do any of these things precisely as described...

AlexWaygood replied:
Hmm... maybe "might" in that case? "Should" doesn't feel sufficiently future-tense-y to me

carljm replied:
But I guess this one we definitely will need, so sure, I'll use "will" :)

carljm commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,430 @@
+
+### Configuration
+
+We should be able to specify non-default red-knot configurations to use in tests, by including a
+TOML fenced code block:
```
Same comment as above; I was preferring to avoid definitive "will" language.

carljm replied:
Ok, looking at it again, I think this is another case where we are pretty committed to doing this.

carljm replied:
I don't think "TOML-fenced" is correct here. The code block is fenced by triple back ticks, it is not fenced by TOML. It is a "fenced code block" whose language is TOML.

AlexWaygood replied:
Ah I see! In that case I think maybe the adjective order is incorrect/confusing -- probably "fenced TOML code block"?

carljm commented:
Ok, pushed an updated that I think addresses all comments.

I also realized in working on the change to require either rule or message-contains, that when a diagnostic doesn't match any assertion we should include its actual column in the "unexpected error" output, otherwise it's hard to debug a failure to match on the column, so I made this change as well. And in the course of making that change, I realized that some tests were not using `dedent` when they should, resulting in silly column numbers, so I fixed that as well.

AlexWaygood commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,447 @@
+
+- `# error: [invalid-assignment]` requires that the matched diagnostic have the rule code
+    `invalid-assignment`. (The square brackets are required.)
+- `# error: "Some text"` requires that the matched diagnostic's full message contain the text `Some text`. (The double quotes are required in the assertion comment; they are not part of the matched
+    text.)
```
nit: some very long lines here!

```suggestion
- `# error: "Some text"` requires that the matched diagnostic's full message contain the text
    `Some text`. (The double quotes are required in the assertion comment; they are not 
    part of the matched text.)
```

carljm replied:
Oops, thanks.

AlexWaygood commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,447 @@
+
+Assertions must contain either a rule code or a contains-text, or both, and may optionally also
+include a column. They must come in order: first column, if present; then rule code, if present;
+then contains-text, if present. For example, an assertion using all three would look like `# error: 8 [invalid-assignment] "Some text"`.
```
```suggestion
then contains-text, if present. For example, an assertion using all three would look like
`# error: 8 [invalid-assignment] "Some text"`.
```

AlexWaygood commented on crates/red_knot_test/README.md:
```diff
@@ -0,0 +1,447 @@
+"Cross-file invalid assignment". The first test involves only a single embedded file, and the second
+test involves two embedded files.
+
+The tests are run independently, in independent in-memory file systems and with new red-knot [Salsa](https://github.com/salsa-rs/salsa)
+databases. This means that each is a from-scratch run of the type checker, with no data persisting from any
+previous test.
```
```suggestion
The tests are run independently, in independent in-memory file systems and with new red-knot
[Salsa](https://github.com/salsa-rs/salsa) databases. This means that each is a
from-scratch run of the type checker, with no data persisting from any previous test.
```

AlexWaygood commented on crates/red_knot_test/README.md:
`````diff
@@ -0,0 +1,447 @@
+```
+````
+
+It should be possible to include a TOML-fenced code block in a single test (as shown), or in a
`````
```suggestion
It should be possible to include a TOML code block in a single test (as shown), or in a
```

AlexWaygood reviewed (approved):
Thanks!

carljm commented:
Thanks for the thorough review!
